src/index.js

```diff
@@ -1,27 +1,46 @@
 import {PureComponent, createElement} from 'react'
 import {create as createJss, getDynamicStyles} from 'jss'
 import preset from 'jss-preset-default'
+
 import filterProps from './utils/filter-props'
+import composeClasses from './utils/compose-classes'
+import type {
+  BaseStylesType,
+  ComponentStylesType,
+  StyledType,
+  StyledElementAttrsType,
+  StyledElementType,
+  TagOrStyledElementTypeype,
+  StyledElementPropsType
+} from './types'
 
 const jssDefault = createJss(preset())
 
-type StyledElementAttrsType = { tag: string, styles: Object }
-type StyledElementType = Function & StyledElementAttrsType
-type tagOrStyledElementTypeype = string | StyledElementType
-type StyledElementPropsType = {
-  classes: Object,
-  children: ?any,
-  className: ?string,
-}
 
-const createStyled = (jss?: Function = jssDefault) => (baseStyles: Object = {}) => {
-  let sheet
-  let dynamicSheet
+const createStyled = (
+  jss?: Function = jssDefault
+) => (
+  baseStyles: BaseStylesType = {}
+): StyledType => {
+  const sheets = {}
   let counter = 0
 
+  const mountSheets = () => {
+    if (!sheets.staticSheet) {
+      sheets.staticSheet = jss.createStyleSheet(baseStyles, {
+        meta: 'StaticBaseSheet',
+      }).attach()
+
+      sheets.dynamicSheet = jss.createStyleSheet({}, {
+        link: true,
+        meta: 'DynamicComponentSheet',
+      }).attach()
+    }
+  }
+
   const styled = (
-    tagOrStyledElement: tagOrStyledElementTypeype,
-    ownStyles: Object
+    tagOrStyledElement: TagOrStyledElementTypeype,
+    ownStyles: ComponentStylesType
   ): StyledElementType => {
     const {tag, styles}: StyledElementAttrsType = typeof tagOrStyledElement === 'string'
```

Comment:
Lets use `tagName` instead of `tag` everywhere, its more correct and aligned with element.tagName